lib/core/src/restore.rs

```diff
@@ -0,0 +1,820 @@
+                    let (lockup_tx_id, claim_tx_id) =
+                        match tx_map.incoming_tx_map.contains_key::<Txid>(first) {
+                            true => (second, first),
+                            false => (first, second),
+                        };
+                    lockup_claim_tx_ids_map.insert(swap_id.clone(), (*lockup_tx_id, *claim_tx_id));
+                }
+                n => {
+                    error!("Script history with unexpected length {n} found while recovering data for Receive Swap {swap_id}")
+                }
+            }
+        }
+
+        info!(
+            "[Recover Receive] Found {} lockup and claim tx pairs from onchain data",
+            lockup_claim_tx_ids_map.len()
+        );
+
+        Ok(RecoveredOnchainDataReceive {
+            lockup_claim_tx_ids: lockup_claim_tx_ids_map,
+        })
+    }
+
+    /// Reconstruct Chain Send Swap tx IDs from the onchain data and the immutable DB data
+    async fn recover_send_chain_swap_tx_ids(
+        &self,
+        tx_map: &TxMap,
+        chain_send_histories_by_swap_id: HashMap<String, SendChainSwapHistory>,
+    ) -> Result<RecoveredOnchainDataChainSend> {
+        let mut lbtc_user_lockup_tx_map: HashMap<String, Txid> = HashMap::new();
+        let mut lbtc_refund_tx_map: HashMap<String, Txid> = HashMap::new();
+        let mut btc_server_lockup_tx_map: HashMap<String, Txid> = HashMap::new();
+        let mut btc_claim_tx_map: HashMap<String, Txid> = HashMap::new();
+
+        for (swap_id, history) in chain_send_histories_by_swap_id {
+            info!("[Recover Chain Send] Checking swap {swap_id}");
+
+            // If a history tx is one of our outgoing txs, it's a lockup tx
+            let maybe_lockup_tx_id = history
+                .lbtc_lockup_script_history
+                .iter()
+                .find(|tx_id| tx_map.outgoing_tx_map.contains_key::<Txid>(tx_id));
+            match maybe_lockup_tx_id {
+                None => {
+                    error!("No lockup tx found when recovering data for Chain Send Swap {swap_id}")
+                }
+                Some(lockup_tx_id) => {
+                    lbtc_user_lockup_tx_map.insert(swap_id.clone(), *lockup_tx_id);
+                }
+            }
+
+            // If a history tx is one of our incoming txs, it's a refund tx
+            let maybe_refund_tx_id = history
+                .lbtc_lockup_script_history
+                .iter()
+                .find(|tx_id| tx_map.incoming_tx_map.contains_key::<Txid>(tx_id));
+            if let Some(refund_tx_id) = maybe_refund_tx_id {
+                lbtc_refund_tx_map.insert(swap_id.clone(), *refund_tx_id);
+            }
+
+            match history.btc_claim_script_history.len() {
+                2 => {
+                    // TODO How to tell the claim tx apart from the lockup tx? Is the order in which they're received from Electrum reliable?
```

Comment:
We might need to fetch these transactions in order to do it reliably.

Reply:
Done, added a call to fetch the BTC txs so we can look at the vouts / vins.